app.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+require 'sinatra'
+require 'sinatra/reloader' if development?
+require_relative './models/memo'
+
+configure do
+  set :haml, escape_html: true
+  set :strict_paths, false
+end
+
+helpers do
+  def find_memo(id)
+    @memo = Memo.find(id)
+  end
+end
+
+get '/' do
+  redirect to('/memos')
+end
+
+get '/memos' do
+  @memos = Memo.all
+
+  haml :'memos/index'
+end
+
+get %r{/memos/(\d+)} do |id|
+  if find_memo(id)
+    haml :'memos/show'
+  else
+    not_found
+  end
+end
+
+get '/memos/new' do
+  haml :'memos/new'
+end
+
+get %r{/memos/(\d+)/edit} do |id|
+  if find_memo(id)
+    haml :'memos/edit'
+  else
+    not_found
+  end
+end
+
+post '/memos' do
+  @memo = Memo.create(title: params[:title], content: params[:content])
+
+  redirect to("/memos/#{@memo.id}")
+end
+
+patch %r{/memos/(\d+)} do |id|
+  if find_memo(id)
+    @memo.update(title: params[:title], content: params[:content])
+
+    redirect to("/memos/#{id}")
+  else
+    not_found
+  end
```

Comment:
この `if` の `then` と `else` は中身が対称ではなくて、`then` が通常の処理、`else` がエラーの処理になっています。処理の重きの置き方は違うけれど、字面上は対称になってしまっている、という感じです。

こういうケースでは、ガード節をつかうと、コードを「前提条件の確認」と「本質的な処理」の2つに分離しつつ、ネストを浅くすることができます（参考: 『リーダブルコード』pp. 91-99）。

```suggestion
  not_found unless find_memo(id)
  
  @memo.update(title: params[:title], content: params[:content])
  redirect to("/memos/#{id}")
```

Sinatra では `not_found` を呼ぶとハンドラーの処理が止まるようです。

> Halt processing and return a 404 Not Found.

https://rubydoc.info/gems/sinatra/Sinatra/Helpers#not_found-instance_method


なので、明示的な脱出は不要そうです。つまり、

```ruby
  unless find_memo(id)
    not_found
    return
  end
```

あるいは

```ruby
  return not_found unless find_memo(id)
```

などとする必要はないようです。ただ、フレームワークによっては、ハンドラーから脱出するところまで自分で書く必要があるものもあるので、注意が必要です。

と、ここまで書いたのですが、 `find_memo` は `@memo` を破壊的に変更しているので、ガードにしてしまうとよくないですね。この件については `def find_memo(id)` のほうにコメントしてみました。

Reply:
なるほど、たしかに「前提条件の確認」と「本質的な処理」は分離して書いた方が良いですね！ガード節を使うとネストが浅くなって読みやすいですね。
ハンドラーから脱出するところまで書く必要があるかどうか注意した方が良いことについても、理解しました。